tiledb/api/src/datatype/mod.rs

```diff
@@ -711,6 +711,256 @@ macro_rules! fn_typed {
     }};
 }
 
+#[macro_export]
+macro_rules! typed_enum {
```

Comment:
Needs comments reflecting the PR description which I will add as this PR is discussed.